src/scales/scale.time.js

```diff
@@ -50,246 +138,271 @@ module.exports = function(Chart) {
-			var parsedData = {
-				labels: [],
-				datasets: []
-			};
-
-			var timestamp;
-
-			helpers.each(chartData.labels, function(label, labelIndex) {
-				var labelMoment = timeHelpers.parseTime(me, label);
-
-				if (labelMoment.isValid()) {
-					// We need to round the time
-					if (timeOpts.round) {
-						labelMoment.startOf(timeOpts.round);
-					}
-
-					timestamp = labelMoment.valueOf();
-					dataMin = Math.min(timestamp, dataMin);
-					dataMax = Math.max(timestamp, dataMax);
-
-					// Store this value for later
-					parsedData.labels[labelIndex] = timestamp;
-				}
-			});
-
-			helpers.each(chartData.datasets, function(dataset, datasetIndex) {
-				var timestamps = [];
+			var chart = me.chart;
+			var options = me.options;
+			var datasets = chart.data.datasets || [];
+			var min = MAX_INTEGER;
+			var max = MIN_INTEGER;
+			var timestamps = [];
+			var labels = [];
+			var i, j, ilen, jlen, data, timestamp;
+
+			// Convert labels to timestamps
+			for (i=0, ilen=chart.data.labels.length; i<ilen; ++i) {
+				timestamp = parse(chart.data.labels[i], me);
+				min = Math.min(min, timestamp);
+				max = Math.max(max, timestamp);
+				labels.push(timestamp);
+			}
 
-				if (typeof dataset.data[0] === 'object' && dataset.data[0] !== null && me.chart.isDatasetVisible(datasetIndex)) {
-					// We have potential point data, so we need to parse this
-					helpers.each(dataset.data, function(value, dataIndex) {
-						var dataMoment = timeHelpers.parseTime(me, me.getRightValue(value));
+			// Convert data to timestamps
+			for (i=0, ilen=datasets.length; i<ilen; ++i) {
+				if (chart.isDatasetVisible(i)) {
+					data = datasets[i].data;
 
-						if (dataMoment.isValid()) {
-							if (timeOpts.round) {
-								dataMoment.startOf(timeOpts.round);
-							}
+					// Let's consider that all data have the same format.
+					if (helpers.isObject(data[0])) {
+						timestamps[i] = [];
 
-							timestamp = dataMoment.valueOf();
-							dataMin = Math.min(timestamp, dataMin);
-							dataMax = Math.max(timestamp, dataMax);
-							timestamps[dataIndex] = timestamp;
+						for (j=0, jlen=data.length; j<jlen; ++j) {
+							timestamp = parse(data[j], me);
+							min = Math.min(min, timestamp);
+							max = Math.max(max, timestamp);
+							timestamps[i][j] = timestamp;
 						}
-					});
+					} else {
+						timestamps[i] = labels.slice();
+					}
 				} else {
-					// We have no x coordinates, so use the ones from the labels
-					timestamps = parsedData.labels.slice();
+					timestamps[i] = [];
 				}
+			}
 
-				parsedData.datasets[datasetIndex] = timestamps;
-			});
-
-			me.dataMin = dataMin;
-			me.dataMax = dataMax;
-			me._parsedData = parsedData;
+			// Enforce limits with user min/max options
+			min = parse(options.time.min, me) || min;
+			max = parse(options.time.max, me) || max;
+
+			// In case there is no valid min/max, let's use today limits
+			min = min === MAX_INTEGER? +moment().startOf('day') : min;
```

Comment:
I'm not familiar with `moment` (might be the first time) and had to search in their docs what `valueOf` actually returns. The unary operator is commonly used to convert string to number (`+'3' === 3`) and I think it's not more confusing than `valueOf` once you know what it does. It may be even more obvious since you would expect a number from `+moment()` which is likely the timestamp. And it's also shorter, so better for minification.